File: backend-node/src/application/useCases/deleteProject.ts
/**
 * Delete Project Use Case
 * 
 * Encapsulates the business logic for deleting a project.
 */

import { ProjectRepository } from '../../infrastructure/repositories/ProjectRepository';

export class DeleteProjectUseCase {
  constructor(private readonly repository: ProjectRepository) {}

  async execute(projectId: string): Promise<void> {
    // The repository's delete method will throw ProjectNotFoundError
    // if the project doesn't exist
    await this.repository.delete(projectId);
  }
}

File: backend-fastapi/app/application/use_cases/delete_project.py
"""
Delete Project Use Case - Application layer.

Encapsulates the business logic for deleting a project.
"""
from uuid import UUID

from app.domain.exceptions import ProjectNotFoundException
from app.infrastructure.repositories.project_repository import ProjectRepository


class DeleteProjectUseCase:
    """
    Use case for deleting a project.
    
    Coordinates the removal of a project from persistence.
    """
    
    def __init__(self, repository: ProjectRepository):
        self.repository = repository
    
    def execute(self, project_id: UUID) -> None:
        """
        Execute the use case.
        
        Args:
            project_id: The UUID of the project to delete.
            
        Raises:
            ProjectNotFoundException: If the project doesn't exist.
        """
        # The repository's delete method will raise ProjectNotFoundException
        # if the project doesn't exist
        self.repository.delete(project_id)
